### DISCUSSION TITLE
[VTM] MapView class not found (android package)

### DISCUSSION THREAD

**Discussion by xamix** [mapsforge/vtm#959] [Q&A] [answered]
Hi,

I'm trying to implement simple MapView
I'm using Kotlin and I have followed the integration guide which lead me to the following Gradle:

```
    implementation 'org.mapsforge:vtm:0.18.0'
    implementation 'org.mapsforge:vtm-android:0.18.0'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-armeabi-v7a'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-arm64-v8a'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86_64'
    implementation 'org.mapsforge:vtm-themes:0.18.0'
    implementation 'net.sf.kxml:kxml2:2.3.0'
    implementation 'org.slf4j:slf4j-api:2.0.0'
    implementation 'com.caverock:androidsvg:1.4'
``` 

And added a MapView in xml:

```
<org.oscim.android.MapView
        android:id="@+id/mapView"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />
```

However it seem it cannot find the package:

```
Class referenced in the layout file, `org.oscim.android.MapView`, was not found in the project or the libraries
```

Even directly in the code it seem it cannot find/import the `org.oscim.android` package where the MapView reside

The gradle sync execute correctly, no error, I have tried to clean cache but nothing helped
I precise some versions here:

Compile SDK: 32
minSdk 28
targetSdk 32
Source and Target compatibility: Java 1.8
IntelliJ 2021.2.1 Patch 1
Gradle Plugin 7.2.1
Gradle Version 7.3.3
plugins {
    id 'com.android.application' version '7.2.1' apply false
    id 'com.android.library' version '7.2.1' apply false
    id 'org.jetbrains.kotlin.android' version '1.7.10' apply false
}

Did I miss a package for get MapView ?

Regards

**Comment by devemux86** [comment]
Android `MapView` is in the vtm-android module.

According to the [integration guide](https://github.com/mapsforge/vtm/blob/master/docs/Integration.md), the dependencies seem correct.
Maybe a project issue with its structure or the dependencies (or internet to download them?).

Besides the [GettingStarted](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java) example, you can also check the other samples in [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example).

**Reply by xamix** [reply]
Hi @devemux86 

I tried by removing the native library in the gradle i.e:

```
    implementation 'org.mapsforge:vtm:0.18.0'
    implementation 'org.mapsforge:vtm-android:0.18.0'
//    implementation 'org.mapsforge:vtm-android:0.18.0:natives-armeabi-v7a'
//    implementation 'org.mapsforge:vtm-android:0.18.0:natives-arm64-v8a'
//    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86'
//    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86_64'
    implementation 'org.mapsforge:vtm-themes:0.18.0'
    implementation 'net.sf.kxml:kxml2:2.3.0'
    implementation 'org.slf4j:slf4j-api:2.0.0'
    implementation 'com.caverock:androidsvg:1.4'
```

And now I can import the MapView but It crash at runtime because it cannot find the native library I presume:

```
Error inflating class org.oscim.android.MapView
 ...
Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: library "libvtm-jni.so" not found
```

So for the moment I do not know how to keep the natives libraries while having the `implementation 'org.mapsforge:vtm-android:0.18.0'` "working"!

If anyone have an idea of what is passing?
For information I do not have the Android NDK, is it mandatory maybe?

Regards

**Reply by xamix** [reply]
By changing order in gradle (put natives before) others, it started to work.
Don't know the mistical behavior behind this as I've tried to close, restart my PC, clean cache before and nothing worked...
Changing order in gradle started to make it work as a magical thing....

**Comment by johnkeel-thork** [comment, marked as answer]
Can confirm the same behaviour on:

Compile Sdk 33
Build Tools 33.0.0
Gradle Plugin 7.3.1
Gradle Version 7.5.1

Seems to started happening only on the most recent grade versions.

The solution is to change the order, just doing this order will fix it:

    implementation 'org.mapsforge:vtm-android:0.18.0:natives-armeabi-v7a'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-arm64-v8a'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86'
    implementation 'org.mapsforge:vtm-android:0.18.0:natives-x86_64'
    implementation 'org.mapsforge:vtm-android:0.18.0'

Natives first!
Probably should add this info in the guides?

**Comment by devemux86** [comment]
In IDEA with Gradle 7.4 and Android Gradle plugin 7.2.2 the following works:
```
implementation 'org.mapsforge:vtm-android:master-SNAPSHOT@jar'
implementation 'org.mapsforge:vtm-android:master-SNAPSHOT:natives-armeabi-v7a@jar'
implementation 'org.mapsforge:vtm-android:master-SNAPSHOT:natives-arm64-v8a@jar'
implementation 'org.mapsforge:vtm-android:master-SNAPSHOT:natives-x86@jar'
implementation 'org.mapsforge:vtm-android:master-SNAPSHOT:natives-x86_64@jar'
```

(similar order to [vtm-android-example/build.gradle](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/build.gradle))
